core/services/ardupilot_manager/firmware_download/lib/FirmwareDownload.py

```diff
@@ -0,0 +1,206 @@
+import enum
+import json
+import pathlib
+import os
+
+from warnings import warn
+
+from packaging.version import Version
+from typing import Any, Dict, List, Optional
+
+
+class Vehicle(enum.Enum):
+    """Valid vehicle types to download"""
+
+    Sub = 1
+    Rover = 2
+    Plane = 3
+    Copter = 4
+
+
+class FirmwareDownload:
+    _manifest_remote = "https://firmware.ardupilot.org/manifest.json.gz"
+    _supported_firmware_format = "apj"
+
+    def __init__(self):
+        self._content = {}
+
+    @staticmethod
+    def _generate_random_filename(length=16) -> pathlib.Path:
```

Comment:
It's safer to create our own random file to avoid any conflict with others programs and services that are running, maybe in the future a service will download a different manifest file and etc, it also does not hurt to be safer. 
With that, we also don't need to track the name of the files or identify the names in the code, we generate names for them and we track this name.

Reply:
Do we even need to download these to the filesystem? At least for the manifest, you could just:
```
from urllib.request import urlopen

with urlopen(URL) as resource:
    versions = json.load(resource)
```

The firmware I assume is going to be loaded from the filesystem when flashing

Reply:
Done